File: services/gateway/src/routes/events.ts
import { Router, Request, Response } from 'express'
import { getPool } from '../db/pool'
import { nudgeBlockchain } from '../utils/nudge'

export const eventsRouter = Router()

// POST /api/events — create a new safety event
eventsRouter.post('/', async (req: Request, res: Response) => {
  const pool = getPool()
  const body = req.body as {
    event_type?: string
    severity?: string
    title?: string
    summary?: string
    lat?: number
    lng?: number
    place_name?: string
    county?: string
    radius_meters?: number
    confidence?: number
    source_count?: number
    source_breakdown?: Record<string, number>
    is_active?: boolean
    started_at?: string
  }

  if (!body.event_type || !body.severity || !body.title || !body.lat || !body.lng || !body.started_at) {
    res.status(400).json({ code: 'VALIDATION_ERROR', message: 'Missing required fields: event_type, severity, title, lat, lng, started_at', retryable: false })
    return
  }

  const client = await pool.connect()
  try {
    await client.query('BEGIN')

    const result = await client.query(
      `INSERT INTO safety_events (
        event_type, severity, title, summary,
        lat, lng, place_name, county, radius_meters,
        confidence, source_count, source_breakdown,
        is_active, started_at
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)
      RETURNING *`,
      [
        body.event_type,
        body.severity,
        body.title,
        body.summary ?? null,
        body.lat,
        body.lng,
        body.place_name ?? null,
        body.county ?? null,
        body.radius_meters ?? 500,
        body.confidence ?? null,
        body.source_count ?? 1,
        JSON.stringify(body.source_breakdown ?? {}),
        body.is_active ?? true,
        body.started_at,
      ]
    )

    const newEvent = result.rows[0]

    if (['AUTHORITATIVE', 'CRITICAL'].includes(body.severity)) {
      await client.query(
        `INSERT INTO publish_jobs (source_type, source_id) VALUES ('SAFETY_EVENT', $1)`,
        [newEvent.id],
      )
    }

    await client.query('COMMIT')
    client.release()

    // nudge is fire-and-forget after commit, not inside transaction
    if (['AUTHORITATIVE', 'CRITICAL'].includes(body.severity)) {
      nudgeBlockchain()
    }

    res.status(201).json(newEvent)
  } catch (err) {
    await client.query('ROLLBACK')
    client.release()
    console.error('POST /api/events error:', err)
    res.status(500).json({ code: 'DB_ERROR', message: 'Could not create event', retryable: true })
  }
})

// GET /api/events?lat&lng&radius_km&severity&type&active_only
eventsRouter.get('/', async (req: Request, res: Response) => {
  const {
    lat, lng,
    radius_km = '10',
    severity,
    type,
    active_only = 'true',
    limit = '50',
  } = req.query

  const pool = getPool()
  const params: unknown[] = []
  const conditions: string[] = []

  if (active_only === 'true') {
    conditions.push('is_active = true')
  }

  if (severity) {
    const severities = String(severity).split(',').map(s => s.trim().toUpperCase())
    params.push(severities)
    conditions.push(`severity = ANY($${params.length}::text[])`)
  }

  if (type) {
    const types = String(type).split(',').map(t => t.trim().toUpperCase())
    params.push(types)
    conditions.push(`event_type = ANY($${params.length}::text[])`)
  }

  // Radius filter using PostGIS earthdistance (km to meters: * 1000)
  if (lat && lng) {
    const latNum = parseFloat(String(lat))
    const lngNum = parseFloat(String(lng))
    const radiusMeters = parseFloat(String(radius_km)) * 1000
    params.push(latNum, lngNum, radiusMeters)
    conditions.push(
      `earth_distance(ll_to_earth($${params.length - 2}, $${params.length - 1}), ll_to_earth(lat, lng)) <= $${params.length}`
    )
  }

  const where = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : ''
  params.push(parseInt(String(limit), 10))

  try {
    const result = await pool.query(
      `SELECT * FROM safety_events ${where} ORDER BY started_at DESC LIMIT $${params.length}`,
      params
    )
    res.json({ events: result.rows, total: result.rowCount })
  } catch (err) {
    console.error('GET /api/events error:', err)
    res.status(500).json({ code: 'DB_ERROR', message: 'Could not fetch events', retryable: true })
  }
})

// GET /api/events/:id
eventsRouter.get('/:id', async (req: Request, res: Response) => {
  const pool = getPool()
  try {
    const result = await pool.query(
      'SELECT * FROM safety_events WHERE id = $1',
      [req.params['id']]
    )
    if (result.rowCount === 0) {
      res.status(404).json({ code: 'NOT_FOUND', message: 'Event not found', retryable: false })
      return
    }
    res.json(result.rows[0])
  } catch (err) {
    console.error('GET /api/events/:id error:', err)
    res.status(500).json({ code: 'DB_ERROR', message: 'Could not fetch event', retryable: true })
  }
})

File: services/gateway/src/routes/events.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::Json,
    routing::get,
    Router,
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{error::AppError, middleware::internal_auth::InternalServiceAuth, AppState};

#[derive(Debug, Clone, Serialize, sqlx::FromRow)]
pub struct SafetyEvent {
    pub id: Uuid,
    pub event_type: String,
    pub severity: String,
    pub title: String,
    pub lat: f64,
    pub lng: f64,
    pub started_at: DateTime<Utc>,
    pub summary: Option<String>,
    pub place_name: Option<String>,
    pub county: Option<String>,
    pub radius_meters: Option<i32>,
    pub confidence: Option<f64>,
    pub source_count: Option<i32>,
    pub source_breakdown: Option<serde_json::Value>,
    pub trust_state: String,
    pub origin_class: String,
    pub is_active: bool,
    pub state: String,
    pub nostr_event_id: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Deserialize)]
pub struct CreateEventBody {
    pub event_type: String,
    pub severity: String,
    pub title: String,
    pub lat: f64,
    pub lng: f64,
    pub started_at: DateTime<Utc>,
    pub summary: Option<String>,
    pub place_name: Option<String>,
    pub county: Option<String>,
    pub radius_meters: Option<i32>,
    pub confidence: Option<f64>,
    pub source_count: Option<i32>,
    pub source_breakdown: Option<serde_json::Value>,
    pub is_active: Option<bool>,
}

#[derive(Deserialize)]
pub struct ListEventsQuery {
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub radius_km: Option<f64>,
    pub severity: Option<String>,
    pub r#type: Option<String>,
    pub active_only: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Deserialize)]
pub struct BoundsQuery {
    pub min_lat: f64,
    pub max_lat: f64,
    pub min_lng: f64,
    pub max_lng: f64,
    pub zoom: Option<f64>,
}

fn viewport_limit(zoom: f64) -> i64 {
    ((1000.0 / (zoom - 8.0).max(1.0)) as i64).clamp(50, 500)
}

async fn create_event(
    State(state): State<AppState>,
    _auth: InternalServiceAuth,
    Json(body): Json<CreateEventBody>,
) -> Result<(StatusCode, Json<SafetyEvent>), AppError> {
    if body.event_type.is_empty() || body.title.is_empty() || body.severity.is_empty() {
        return Err(AppError::BadRequest(
            "event_type, title, severity are required".into(),
        ));
    }

    if !(-90.0..=90.0).contains(&body.lat) || !(-180.0..=180.0).contains(&body.lng) {
        return Err(AppError::BadRequest(
            "lat must be in -90..90, lng in -180..180".into(),
        ));
    }

    let event = sqlx::query_as::<_, SafetyEvent>(
        "INSERT INTO safety_events
           (event_type, severity, title, lat, lng, started_at, summary, place_name, county,
            radius_meters, confidence, source_count, source_breakdown, is_active)
         VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)
         RETURNING id, event_type, severity, title, lat, lng, started_at,
                   summary, place_name, county, radius_meters, confidence,
                   source_count, source_breakdown, trust_state, origin_class,
                   is_active, state,
                    nostr_event_id, created_at, updated_at",
    )
    .bind(&body.event_type)
    .bind(&body.severity)
    .bind(&body.title)
    .bind(body.lat)
    .bind(body.lng)
    .bind(body.started_at)
    .bind(&body.summary)
    .bind(&body.place_name)
    .bind(&body.county)
    .bind(body.radius_meters)
    .bind(body.confidence)
    .bind(body.source_count)
    .bind(&body.source_breakdown)
    .bind(body.is_active.unwrap_or(true))
    .fetch_one(&state.db)
    .await?;

    let msg = serde_json::json!({ "type": "NEW_EVENT", "payload": event });
    state.hub.broadcast(
        body.county.as_deref(),
        serde_json::to_string(&msg).unwrap().into(),
    );
    let viewport_json = serde_json::json!({
        "id": event.id,
        "event_type": event.event_type,
        "severity": event.severity,
        "state": event.state,
        "trust_state": event.trust_state,
        "title": event.title,
        "lat": event.lat,
        "lng": event.lng,
        "started_at": event.started_at,
    });
    let _ = state.event_tx.send(crate::ws::ViewportEvent {
        id: event.id,
        lat: event.lat,
        lng: event.lng,
        severity: event.severity.clone(),
        event_json: viewport_json.to_string().into(),
    });

    Ok((StatusCode::CREATED, Json(event)))
}

async fn list_events(
    State(state): State<AppState>,
    Query(q): Query<ListEventsQuery>,
) -> Result<Json<serde_json::Value>, AppError> {
    let radius_m = q.radius_km.unwrap_or(10.0) * 1000.0;
    let limit = q.limit.unwrap_or(50).min(200);
    let active_only = q.active_only.as_deref() != Some("false");

    let severity_filter: Option<Vec<String>> = q
        .severity
        .as_deref()
        .map(|s| s.split(',').map(|x| x.trim().to_uppercase()).collect());
    let type_filter: Option<Vec<String>> = q
        .r#type
        .as_deref()
        .map(|s| s.split(',').map(|x| x.trim().to_uppercase()).collect());

    let events = sqlx::query_as::<_, SafetyEvent>(
        "SELECT id, event_type, severity, title, lat, lng, started_at,
                summary, place_name, county, radius_meters, confidence,
                source_count, source_breakdown, trust_state, origin_class,
                is_active, state,
                 nostr_event_id, created_at, updated_at
         FROM safety_events
         WHERE ($1::float8 IS NULL OR
                earth_distance(ll_to_earth($1,$2), ll_to_earth(lat,lng)) <= $3)
           AND ($4::text[] IS NULL OR severity = ANY($4))
           AND ($5::text[] IS NULL OR event_type = ANY($5))
           AND (NOT $6 OR is_active = true)
         ORDER BY created_at DESC
         LIMIT $7",
    )
    .bind(q.lat)
    .bind(q.lng)
    .bind(radius_m)
    .bind(severity_filter)
    .bind(type_filter)
    .bind(active_only)
    .bind(limit)
    .fetch_all(&state.db)
    .await?;

    let total = events.len() as i64;
    Ok(Json(
        serde_json::json!({ "events": events, "total": total }),
    ))
}

async fn get_event(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<SafetyEvent>, AppError> {
    let event = sqlx::query_as::<_, SafetyEvent>(
        "SELECT id, event_type, severity, title, lat, lng, started_at,
                summary, place_name, county, radius_meters, confidence,
                source_count, source_breakdown, trust_state, origin_class,
                is_active, state,
                 nostr_event_id, created_at, updated_at
         FROM safety_events WHERE id = $1",
    )
    .bind(id)
    .fetch_optional(&state.db)
    .await?
    .ok_or(AppError::NotFound)?;

    Ok(Json(event))
}

async fn list_events_by_bounds(
    State(state): State<AppState>,
    Query(q): Query<BoundsQuery>,
) -> Result<Json<serde_json::Value>, AppError> {
    let limit = viewport_limit(q.zoom.unwrap_or(12.0));
    let events = sqlx::query_as::<_, SafetyEvent>(
        "SELECT id, event_type, severity, title, lat, lng, started_at,
                summary, place_name, county, radius_meters, confidence,
                source_count, source_breakdown, trust_state, origin_class,
                is_active, state,
                 nostr_event_id, created_at, updated_at
           FROM safety_events
          WHERE geog && ST_MakeEnvelope($1, $2, $3, $4, 4326)::geography
            AND state NOT IN ('RESOLVED', 'EXPIRED')
          ORDER BY created_at DESC
          LIMIT $5",
    )
    .bind(q.min_lng)
    .bind(q.min_lat)
    .bind(q.max_lng)
    .bind(q.max_lat)
    .bind(limit)
    .fetch_all(&state.db)
    .await?;
    let total = events.len() as i64;
    Ok(Json(
        serde_json::json!({ "events": events, "total": total }),
    ))
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_events).post(create_event))
        .route("/bounds", get(list_events_by_bounds))
        .route("/:id", get(get_event))
}

impl From<SafetyEvent> for sentinel_core::Event {
    fn from(row: SafetyEvent) -> Self {
        Self {
            id: row.id,
            event_type: row.event_type,
            severity: row.severity,
            title: row.title,
            lat: row.lat,
            lng: row.lng,
            started_at: row.started_at,
            summary: row.summary,
            place_name: row.place_name,
            county: row.county,
            is_active: row.is_active,
            state: Some(row.state),
            created_at: row.created_at,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use uuid::Uuid;

    fn sample_db_row() -> SafetyEvent {
        let now = Utc::now();
        SafetyEvent {
            id: Uuid::nil(),
            event_type: "FLOOD".into(),
            severity: "CRITICAL".into(),
            title: "Flooding downtown".into(),
            lat: -1.2921,
            lng: 36.8219,
            started_at: now,
            summary: Some("Roads impassable".into()),
            place_name: Some("CBD".into()),
            county: Some("Nairobi".into()),
            radius_meters: Some(2000),
            confidence: Some(0.95),
            source_count: Some(3),
            source_breakdown: None,
            trust_state: "confirmed".into(),
            origin_class: "machine".into(),
            is_active: true,
            state: "ACTIVE".into(),
            nostr_event_id: Some("nevent1abc".into()),
            created_at: now,
            updated_at: now,
        }
    }

    #[test]
    fn from_safety_event_maps_core_fields() {
        let row = sample_db_row();
        let e = sentinel_core::Event::from(row);
        assert_eq!(e.id, Uuid::nil());
        assert_eq!(e.event_type, "FLOOD");
        assert_eq!(e.severity, "CRITICAL");
        assert_eq!(e.lat, -1.2921);
        assert_eq!(e.county, Some("Nairobi".into()));
        assert!(e.is_active);
    }

    #[test]
    fn from_safety_event_drops_db_only_fields() {
        let row = sample_db_row();
        let e = sentinel_core::Event::from(row);
        // These fields exist on the domain type (compile-time check):
        let _: Uuid = e.id;
        let _: String = e.event_type;
        let _: String = e.severity;
        let _: bool = e.is_active;
    }

    #[test]
    fn safety_event_api_exposes_only_current_fields() {
        let value = serde_json::to_value(sample_db_row()).unwrap();
        assert!(value.get("nostr_event_id").is_some());
        assert!(value.get("bitcoin_txid").is_none());
        assert!(value.get("bitcoin_block").is_none());
        assert!(value.get("anchor_hash").is_none());
    }

    #[test]
    fn viewport_limit_clamped_to_max_at_low_zoom() {
        assert_eq!(viewport_limit(8.0), 500);
    }

    #[test]
    fn viewport_limit_mid_zoom() {
        assert_eq!(viewport_limit(12.0), 250);
    }

    #[test]
    fn viewport_limit_clamped_to_min_at_high_zoom() {
        assert_eq!(viewport_limit(30.0), 50);
    }
}
